modules/tinymce/src/core/main/ts/fmt/FormatChanged.ts

```diff
@@ -5,101 +5,167 @@
-    Tools.each(parents, (node: Node) => {
-      if (editor.formatter.matchNode(node, format, {}, data.similar)) {
-        if (formatsList.indexOf(format) === -1) {
-          // Execute callbacks
-          Arr.each(data.callbacks, (callback: FormatChangeCallback) => {
-            callback(true, { node, format, parents });
-          });
-
-          newFormats[format] = data.callbacks;
-        }
 
-        matchedFormats[format] = data.callbacks;
-        return false;
-      }
+  editor.on('FormatApply FormatRemove', (e: EditorEvent<FormatEvent>) => {
+    const element = editor.selection.getStart();
+    updateAndFireChangeCallbacks(editor, element, registeredFormatListeners.get());
+  });
+};
 
-      if (MatchFormat.matchesUnInheritedFormatSelector(editor, node, format)) {
+const matchingNode = (editor: Editor, parents: Element[], format: string, similar: boolean, vars?: FormatVars): Optional<Element> => {
+  const isMatchingNode = (node: Element) => {
+    const matchingFormat = editor.formatter.matchNode(node, format, vars ?? {}, similar);
+    return !Type.isUndefined(matchingFormat);
+  };
+  const isUnableToMatch = (node: Element) => {
+    if (MatchFormat.matchesUnInheritedFormatSelector(editor, node, format)) {
+      return true;
+    } else {
+      if (!similar) {
+        // If we want to find an exact match, then finding a similar match halfway up the parents tree is bad
+        return Type.isNonNullable(editor.formatter.matchNode(node, format, vars, true));
+      } else {
         return false;
       }
-    });
-  });
-
-  // Check if current formats still match
-  const remainingFormats = filterRemainingFormats(currentFormats.get(), matchedFormats, elm, parents);
-
-  // Update the current formats
-  currentFormats.set({
-    ...newFormats,
-    ...remainingFormats
-  });
+    }
+  };
+  return Arr.findUntil(parents, isMatchingNode, isUnableToMatch);
 };
 
-const filterRemainingFormats = (currentFormats: FormatCallbacks, matchedFormats: FormatCallbacks, elm: Element, parents: Node[]) => Obj.bifilter(currentFormats, (callbacks: FormatChangeCallback[], format: string) => {
-  if (!Obj.has(matchedFormats, format)) {
-    // Execute callbacks
-    Arr.each(callbacks, (callback: FormatChangeCallback) => {
-      callback(false, { node: elm, format, parents });
-    });
+const getParents = (editor: Editor, elm?: Element): Element[] => {
+  const element = elm ?? editor.selection.getNode();
+  return Arr.filter(FormatUtils.getParents(editor.dom, element), (node) =>
+    NodeType.isElement(node) && !node.getAttribute('data-mce-bogus')
+  );
+};
 
+const hasChanged = (state: Cell<boolean>, newState: boolean): boolean => {
+  if (state.get() === newState) {
     return false;
   } else {
+    state.set(newState);
     return true;
   }
-}).t;
+};
+
+const updateAndFireChangeCallbacks = (
+  editor: Editor,
+  elm: Element,
+  registeredCallbacks: RegisteredFormats
+) => {
```

Comment:
```suggestion
const updateAndFireChangeCallbacks = (editor: Editor, elm: Element, registeredCallbacks: RegisteredFormats) => {
```
This shouldn't be wrapped, as it's well below the 160 char limit.